### PR TITLE
1881 HyperConformance enceladus_info_version from payload 

### LINKED ISSUE #1881: HyperConformance report version from payload
## Background
Currently, in the [HyperConformance](https://github.com/AbsaOSS/enceladus/blob/develop/spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala) plugin, the `enceladus_info_version` can only be set through `report.version` with a static value. It is a desired feature from source teams to set `enceladus_info_version` dynamically based on a field from the payload, similarly how `event.timestamp.column` can be used to set the value for `enceladus_info_date`.

## Feature
Add an attribute (e.g. `report.version.column`) that accepts the name of an Int-column. That Int-column should be used as the value for `enceladus_info_version`. 

## Example
Given a dataframe

| key | value | custom_version |
| --- | --- | --- |
| 100 | abc | 1 |
| 101 | xyz | 1 |
| 100 | ABC | 2 |
| 101 | XYZ | 2 | 

With HyperConformance configuration

- `dataset.name=Some_Dataset`
- `dataset.version=1`
-  ...
- `report.version.column=custom_version`

The conformed dataframe should look like this
| key | value | custom_version | enceladus_info_date | enceladus_info_date_string | enceladus_info_version |
| --- | --- | --- | --- | --- | --- |
| 100 | abc | 1 | ... | ... | 1 |
| 101 | xyz | 1 | ... | ... | 1 |
| 100 | ABC | 2 | ... | ... | 2 |
| 101 | XYZ | 2 | ... | ... | 2 |

(Dates are omitted)

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #1881

### CODE REVIEW

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+object InfoVersionFactory {
+  import za.co.absa.enceladus.conformance.HyperConformanceAttributes._
+
+  def getFactoryFromConfig(conf: Configuration): InfoVersionFactory = {
+    if (conf.containsKey(reportVersionKey)) {
+      val reportVersion = conf.getInt(reportVersionKey)
+      log.info(s"Info version: configuration reportVersion = $reportVersion")
+      new InfoVersionLiteralFactory(reportVersion)
+    } else if (conf.containsKey(reportVersionColumnKey)) {
```
Comment (kevinwallimann) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:43]: In https://github.com/AbsaOSS/enceladus/blob/c96f779ade64715f26309def34eee40c99507b22/spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoDateFactory.scala#L57-L58, if the column does not exist, an `IllegalArgumentException` is thrown. I'd suggest doing the same here

**Review by kevinwallimann — CHANGES_REQUESTED** [review]

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+class InfoVersionColumnFactory(columnName: String) extends InfoVersionFactory {
+  override def getInfoVersionColumn(df: DataFrame): Column = {
+    val dt: Option[StructField] = SchemaUtils.getField(columnName, df.schema)
+    dt match {
+      case Some(_) => col(columnName)
+      case None => throw new IllegalArgumentException(s"The specified info column does not exist: $columnName")
+    }
+  }
+}
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:41]: I would place these two within `InfoVersionFactory` and mark them `private`.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  def getFactoryFromConfig(conf: Configuration): InfoVersionFactory = {
+    if (conf.containsKey(reportVersionKey)) {
+      val reportVersion = conf.getInt(reportVersionKey)
+      log.info(s"Info version: configuration reportVersion = $reportVersion")
+      new InfoVersionLiteralFactory(reportVersion)
+    } else if (conf.containsKey(reportVersionColumnKey)) {
+      val infoVersionColumn = conf.getString(reportVersionColumnKey)
+      log.info(s"Info version: configuration infoVersionColumn = $infoVersionColumn")
+      new InfoVersionColumnFactory(infoVersionColumn)
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:54]: What if both `reportVersionKey` and `reportVersionColumnKey` are defined? Is the behavior that  `reportVersionKey`  takes precedence the desired one? @kevinwallimann

**Review by kevinwallimann — APPROVED** [review]

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala (hunk 3 above)
Comment (kevinwallimann) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:54]: Hi @benedeki Good question. Yes that behaviour is the desired one. Tbh, for me both options make equally sense or no sense, but this behaviour is consistent with https://github.com/AbsaOSS/enceladus/blob/develop/spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoDateFactory.scala#L74-L87

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  }
+}
+
+object InfoVersionFactory {
+  import za.co.absa.enceladus.conformance.HyperConformanceAttributes._
+
+  def getFactoryFromConfig(conf: Configuration): InfoVersionFactory = {
+    if (conf.containsKey(reportVersionKey)) {
+      val reportVersion = conf.getInt(reportVersionKey)
+      log.info(s"Info version: configuration reportVersion = $reportVersion")
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:49]: Would make it in similar to the info date, no class-like words:
```suggestion
      log.info(s"Information version: Explicit from the job configuration = $reportVersion")
```

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      new InfoVersionLiteralFactory(reportVersion)
+    } else if (conf.containsKey(reportVersionColumnKey)) {
+      val infoVersionColumn = conf.getString(reportVersionColumnKey)
+      log.info(s"Info version: configuration infoVersionColumn = $infoVersionColumn")
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/streaming/InfoVersionFactory.scala:53]: Same with the language of the log line as above
```suggestion
      log.info(s"Information version: Derived from the configured column = $infoVersionColumn")
```

**Review by benedeki — CHANGES_REQUESTED** [review]
Functionally no issue, the few style and languages changes please.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala [hunk 6]
```diff
@@ -65,7 +65,7 @@ class HyperConformance (implicit cmd: ConformanceConfig,
     import za.co.absa.enceladus.utils.implicits.DataFrameImplicits.DataFrameEnhancements
 
     val infoDateColumn = infoDateFactory.getInfoDateColumn(rawDf)
-    val infoVersionColumn = infoVersionFactory.getInfoVersionColumn(rawDf)
+    val infoVersionCol = infoVersionFactory.getInfoVersionColumn(rawDf)
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/HyperConformance.scala:68]: I think you changed at wrong place. Here it was consistent, the inconsistency was on line 146 😉

**Review by dk1844 — DISMISSED** [review]
Otherwise LGTM (just read the code).

**Review by dk1844 — DISMISSED** [review]
LGTM now.

**Review by benedeki — APPROVED** [review]
- [x] code reviewed
- [x] pulled
- [ ] built
- [ ] run

**Pull request merged**